### ISSUE TITLE
Using top-up scans to resolve spatial distortions

### ISSUE THREAD

**Issue by jawaltz** [nipreps/fmriprep#2407]
We have acquired top-up scans, for purpose of resolving spatial distortions resulting from A --> P acquisitions. I have been shown the distortions, but I have never corrected for them myself. I have used fmriprep to preprocess all scans acquired in the A --> P directions.

There are algorithms for distortion correction with blip-up-blip-down data, e.g.:

https://www.fmrib.ox.ac.uk/primers/intro_primer/ExBox20/IntroBox20.html

Will this be added to the pipeline? Is it already PART of the pipeline? Can I use these scans in a post-hoc manner, in preprocessing? 

Thanks very much,

Jim Waltz

**Comment by julfou81** [comment]
Hi,

Distorsion correction is part of fmriprep processing if you specify correctly which scans should be use for distorsion correction.
To do so you need to:
- put your topup scans (a pair of volumes with A->P and P->A phase encoding directions), with extension _epi.nii.gz, in a BIDS folder named /fmap.
- For each topup volume, have a son file where you have a field IntendedFor which specify which fMRI run will need to be distorsion corrected with these topup scans (correction named pepolar distortion correction in fmriprep documentation). 
Exemple: ```  "IntendedFor": [
    "func/sub-01_task-audiodisc_run-01_bold.nii.gz",
    "func/sub-01_task-audiodisc_run-02_bold.nii.gz",
    "func/sub-01_task-audiodisc_run-03_bold.nii.gz",
    "func/sub-01_task-audiodisc_run-04_bold.nii.gz",
    "func/sub-01_task-audiodisc_run-05_bold.nii.gz",
    "func/sub-01_task-audiodisc_run-06_bold.nii.gz"
  ]```

Then, when you run fmriprep on your dataset, the fMRI runs will be corrrected by the 'pepolar' method (which is not actually FSL-topup but rather AFNI's 3dQwarp ).

**Comment by jawaltz** [comment]
That is very helpful. Thank you.

So, I just need to put the top-up scans in their proper place, and edit the json file to say what they should be paired with?

Do I do the same for the single-band reference files? So, 6 runs = 6 multi-band top-up scans + 6 single-band reference scans, referring to the 6 multi-band scan sequences + 6 single-band reference scans acquired in the A-->P direction?

**Comment by julfou81** [comment]
Yes, for the topup scans, you should copy the AP/PA pair in the fmap folder along with the completed json files.
The single band reference scans of your functional runs in AP (_sbref.nii.gz)  will be automatically used as target for motion correction and registration as long as their names is matching the bold runs and that you don't use the ```--ignore sbref``` flag in the fmriprep command. No need to edit the json file for those.

Regarding the topup scans, you raise an interesting point. You could create a topup pair for each runs in 2 ways: 
- either copy each the matching pair of SBref scan in AP and PA for each run in the fmap/ folder
-  or put the PA volume (not SBref) of the topup scan and extract and copy also the closest (in time) volume from the AP run (i.e. 1st or last volume of the AP run) that you put in the fmap/ folder.

Both options are valid I think and should provide similar results. 

Since the SBref images have a better SNR and GM/WM contrast than the rest of the other EPI volumes, they may be better candidates to form a topup pair... but I am not sure that one would really see a difference between the two options.. it should be tested I guess.
In our center , we use SE-EPI scans to form the topup pair, following what is doing the HCP.

**Comment by jawaltz** [comment]
So, the top-up scans always need to come in pairs?

I think I had just been putting the AP SBRef files in the func folders. Do I now put them in the fmap folders, or do I put them in BOTH places?

I think grabbing the last volume from the 426-volume multi-band runs would be too much of a pain.

Thanks,

jw

**Comment by julfou81** [comment]
It is a good remark, it is what I understood, but maybe I am wrong. Besides, in fmriprep it is not topup that is used but AFNI 3dQwarp, one should look at how it is implemented in there. I did try the option I told you with grabbing the volume of the AP serie and copying in in the fmap/ folder to form a topup pair and fmriprep ran smoothly with that. But I didn't try with only one opposite  PA volume in fmap. I will look further and post in this thread what I discover, it is something that interests me also.

**Comment by julfou81** [comment]
I do think that you need to provide the pair of opposite phase encoding direction epi image for what is called the pepolar correction method.

Here is what is said in the documentation:
https://fmriprep.org/en/stable/sdc.html
Phase Encoding POLARity (**PEPOLAR**; also called blip-up/blip-down; init_pepolar_unwarp_wf()): acquire at least two images with varying PE directions. Hence, the realization of distortion is different between the different acquisitions. The displacements map is estimated with an image registration process between the different PE acquisitions, regularized by the readout time.

and in the code:
https://www.nipreps.org/sdcflows/api/sdcflows.workflows.fit.pepolar.html
**sdcflows.workflows.fit.pepolar.init_3dQwarp_wf**(omp_nthreads=1, debug=False, name='pepolar_estimate_wf')[source]
Create the PEPOLAR field estimation workflow based on AFNI’s 3dQwarp.

This workflow takes in two EPI files that MUST have opposed PE direction. Therefore, EPIs with orthogonal PE directions are not supported.

**Comment by julfou81** [comment]
You could write a script for the extraction/copy of the last volume of each run:
in bash:

```bold='name_of_your_bold_run'
nvol=`fslval $bold dim4` # number of volumes in the bold run
fslroi $bold ../fmap/sub-XX_dir-AP_run-1_epi $(expo $nvol - 1) 1 # the last volume of the run is extracted and copied in fmap/```

**Comment by jawaltz** [comment]
I had been specifying sbref data sets with the 'sbref' extension, and I had been specifying the direction of acquisition as '_acq-AP_' and '_acq-PA_'. 

Has the bids convention changed, such that I should specify sbref data sets with the 'epi' extension, and the direction of acquisition with '_dir-AP_' and '_dir-PA_'? At least for fieldmaps?

fmriprep is no longer happy with file names it was previously ok with. But that could be for several reasons. I am, for example, some versions behind:

You are using fMRIPrep-20.0.2, and a newer version of fMRIPrep is available: 20.2.1.
Please check out our documentation about how and when to upgrade:
https://fmriprep.readthedocs.io/en/latest/faq.html#upgrading
WARNING: Version 20.0.2 of fMRIPrep (current) has been FLAGGED
(reason: fsLR resampling error (problematic only when using the --cifti-output flag); Phase-difference fieldmap bug).
That means some severe flaw was found in it and we strongly
discourage its usage.
Making sure the input data is BIDS compliant (warnings can be ignored in most cases).
bids-validator@1.4.0

	1: [ERR] Files with such naming scheme are not part of BIDS specification. This error is most commonly caused by typos in file names that make them not BIDS compatible. Please consult the specification and make sure your files are named correctly. If this is not a file naming issue (for example when including files not yet covered by the BIDS specification) you should include a ".bidsignore" file in your dataset (see https://github.com/bids-standard/bids-validator#bidsignore for details). Please note that derived (processed) data should be placed in /derivatives folder and source data (such as DICOMS or behavioural logs in proprietary formats) should be placed in the /sourcedata folder. (code: 1 - NOT_INCLUDED)
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-01_sbref.json
			Evidence: sub-M083801_task-CBT_acq-AP_run-01_sbref.json
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-01_sbref.nii
			Evidence: sub-M083801_task-CBT_acq-AP_run-01_sbref.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-02_sbref.json
			Evidence: sub-M083801_task-CBT_acq-AP_run-02_sbref.json
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-02_sbref.nii
			Evidence: sub-M083801_task-CBT_acq-AP_run-02_sbref.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-03_sbref.json
			Evidence: sub-M083801_task-CBT_acq-AP_run-03_sbref.json
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-03_sbref.nii
			Evidence: sub-M083801_task-CBT_acq-AP_run-03_sbref.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-04_sbref.json
			Evidence: sub-M083801_task-CBT_acq-AP_run-04_sbref.json
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-04_sbref.nii
			Evidence: sub-M083801_task-CBT_acq-AP_run-04_sbref.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-05_sbref.json
			Evidence: sub-M083801_task-CBT_acq-AP_run-05_sbref.json
		./sub-M083801/fmap/sub-M083801_task-CBT_acq-AP_run-05_sbref.nii
			Evidence: sub-M083801_task-CBT_acq-AP_run-05_sbref.nii

**Comment by julfou81** [comment]
Yes, I came across this as well. The BIDS specification moved along the way. For PEPOLAR correction, the epi scans in the 'fmap' folder must have the extension _epi. However in the 'func' folder, the SBref scans still have the _sbref extension. The _acq- flag instead of the _dir- is not an error I think but I did change all my _acq-AP to _dir-AP on my datasets to better follow the new BIDS specification.
Details are here: 
https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#case-4-multiple-phase-encoded-directions-pepolar

**Comment by julfou81** [comment]
The dir- label follows the same logic for dwi scans as well.

**Comment by jawaltz** [comment]
I renamed my files, but I am still having no luck.

I see that the dir-<label> entity is REQUIRED for these files and that this key-value pair MUST be used in addition to the REQUIRED PhaseEncodingDirection metadata field.

That is all true for my data now, but fmriprep is still not happy:

---

Making sure the input data is BIDS compliant (warnings can be ignored in most cases).
bids-validator@1.4.0

	1: [ERR] Files with such naming scheme are not part of BIDS specification. This error is most commonly caused by typos in file names that make them not BIDS compatible. Please consult the specification and make sure your files are named correctly. If this is not a file naming issue (for example when including files not yet covered by the BIDS specification) you should include a ".bidsignore" file in your dataset (see https://github.com/bids-standard/bids-validator#bidsignore for details). Please note that derived (processed) data should be placed in /derivatives folder and source data (such as DICOMS or behavioural logs in proprietary formats) should be placed in the /sourcedata folder. (code: 1 - NOT_INCLUDED)
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-01_epi.json
			Evidence: sub-M083801_task-CBT_dir-AP_run-01_epi.json
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-01_epi.nii
			Evidence: sub-M083801_task-CBT_dir-AP_run-01_epi.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-02_epi.json
			Evidence: sub-M083801_task-CBT_dir-AP_run-02_epi.json
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-02_epi.nii
			Evidence: sub-M083801_task-CBT_dir-AP_run-02_epi.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-03_epi.json
			Evidence: sub-M083801_task-CBT_dir-AP_run-03_epi.json
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-03_epi.nii
			Evidence: sub-M083801_task-CBT_dir-AP_run-03_epi.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-04_epi.json
			Evidence: sub-M083801_task-CBT_dir-AP_run-04_epi.json
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-04_epi.nii
			Evidence: sub-M083801_task-CBT_dir-AP_run-04_epi.nii
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-05_epi.json
			Evidence: sub-M083801_task-CBT_dir-AP_run-05_epi.json
		./sub-M083801/fmap/sub-M083801_task-CBT_dir-AP_run-05_epi.nii
			Evidence: sub-M083801_task-CBT_dir-AP_run-05_epi.nii
		... and 14 more files having this issue (Use --verbose to see them all).

**Comment by julfou81** [comment]
I think the task- label should not be used in fmap/

**Comment by jawaltz** [comment]
Ok, it is now running.

Just to be sure: the "IntendedFor" lines in BOTH run-specific json files in the fmap folder 

sub-M083801_dir-AP_run-01_epi.json
sub-M083801_dir-PA_run-01_epi.json

should specify the same bold file ("sub-M083801_task-CBT_acq-AP_run-01_bold.nii")?

**Comment by julfou81** [comment]
yes it should be ```"IntendedFor": [ "func/sub-M083801_task-CBT_acq-AP_run-01_bold.nii"]```for both json files in fmap/

**Comment by oesteban** [comment]
@julfou81 @jawaltz yes, we've recently moved on with the BIDS specs precisely because the former specs could not account for use cases like the one @jawaltz is proposing with sufficient unambiguity. Please check carefully the new specs here: https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#echo-planar-imaging-and-b0-mapping - particularly the new `B0FieldIndentifier` and `B0FieldSources` metadata entries.

However, the new BIDS specifications are very recent and supported only by the bleeding-edge version of our component for susceptibility-derived distortions, [*SDCFlows*](https://github.com/nipreps/sdcflows). The latest stable release of *fMRIPrep* does NOT use this newest *SDCFlows*. This is to say that the new BIDS specs are not yet supported (although @mgxd is working on it (check progress on #2392). Hard to see a timeline at this moment for a new *fMRIPrep* release with it.

Another positive aspect of the future *SDCFlows* component is that estimation is indeed done with FSL TOPUP

**Comment by jawaltz** [comment]
Hi-

The script is running now, but since we installed fmriprep-20.2.1.simg, we get the following error

---
ModuleNotFoundError: No module named 'niworkflows.interfaces.ants'
---

which leads to other errors, like

---
210507-11:00:34,623 nipype.utils WARNING:
	 No metadata was found in the pkl file. Make sure you are currently using the same Nipype version from the generated pkl.
210507-11:00:34,635 nipype.workflow ERROR:
	 Node inu_n4 failed to run on host mprc00.cluster.local.
210507-11:00:34,647 nipype.workflow ERROR:
	 Saving crash info to /ana/fmriprep/sub-M083801/log/20210507-105842_5532f9d1-6c45-47e8-beb6-a7a49b2ac0cc/crash-20210507-110034-jwaltz-inu_n4-776052f5-b074-4b1e-b68f-57120eabbb41.txt
---

Do you know what I need to do?

Thanks,

jw

**Comment by jawaltz** [comment]
Hi Esteban-

Thanks so much. I emptied out the work directory, but I think I have two more small problems:
- My default .cache directory is small; and
- I don't think my PYTHONPATH is set.

In the linked solution, you say to unset PYTHONPATH, and export PYTHONNOUSERSITE=1.

When I do those things, it returns nothing. But "which python" returns '/usr/bin/python/', which points to 'python2', which points to 'python2.7'. Is that an issue?

Also, can I sent the cache directory, in fmriprep?

**Comment by oesteban** [comment]
Hi, Python 2 is very outdated - your environment is not adequate to run MRIQC.

Could you perhaps try Docker or singularity?
